We follow the suggesstions to support the mentioned data dtypes of float16 and blfoat16, but just adding support in `check_variable_and_dtype` in `exp` and `log` op did not work. 

### What we did?
1. rebase our branch on the latest develop branch
2. add supported data types in `log` and `exp ` op
3. fix our code to support these data types
4. compile this version and test the newly supported datatypes.

### What we find?
Our unittests on these newly supported data types failed both in `float16` and `bfloat16`, the error messages are pasted as following:

733: test_poisson_nll_loss failed
733:  ........E.................
733: ======================================================================
733: ERROR: test_api (test_poisson_nll_loss.TestPoissonNLLLossFloat16Case)
733: ----------------------------------------------------------------------
733: Traceback (most recent call last):
733:   File "/paddle/build/python/paddle/fluid/tests/unittests/test_poisson_nll_loss.py", line 205, in test_api
733:     self.test_static_case(dtype="float16")
733:   File "/paddle/build/python/paddle/fluid/tests/unittests/test_poisson_nll_loss.py", line 113, in test_static_case
733:     fetch_list=[out1, out2],
733:   File "/paddle/build/python/paddle/fluid/executor.py", line 1458, in run
733:     use_prune=use_prune,
733:   File "/paddle/build/python/paddle/fluid/executor.py", line 1660, in _run_impl
733:     scope, list(feed.keys()), fetch_list, return_numpy
733:   File "/paddle/build/python/paddle/fluid/executor.py", line 683, in run
733:     scope, feed_names, fetch_list
733: NotImplementedError: In user code:
733: 
733:     File "/paddle/tools/test_runner.py", line 70, in <module>
733:       main()
733:     File "/paddle/tools/test_runner.py", line 52, in main
733:       res = unittest.TextTestRunner(stream=buffer).run(tests)
733:     File "/usr/lib/python3.7/unittest/runner.py", line 176, in run
733:       test(result)
733:     File "/usr/lib/python3.7/unittest/suite.py", line 84, in __call__
733:       return self.run(*args, **kwds)
733:     File "/usr/lib/python3.7/unittest/suite.py", line 122, in run
733:       test(result)
733:     File "/usr/lib/python3.7/unittest/suite.py", line 84, in __call__
733:       return self.run(*args, **kwds)
733:     File "/usr/lib/python3.7/unittest/suite.py", line 122, in run
733:       test(result)
733:     File "/usr/lib/python3.7/unittest/case.py", line 676, in __call__
733:       return self.run(*args, **kwds)
733:     File "/usr/lib/python3.7/unittest/case.py", line 628, in run
733:       testMethod()
733:     File "/paddle/build/python/paddle/fluid/tests/unittests/test_poisson_nll_loss.py", line 205, in test_api
733:       self.test_static_case(dtype="float16")
733:     File "/paddle/build/python/paddle/fluid/tests/unittests/test_poisson_nll_loss.py", line 99, in test_static_case
733:       reduction=reduction,
733:     File "/paddle/build/python/paddle/nn/functional/loss.py", line 1569, in poisson_nll_loss
733:       loss_out = paddle.exp(input) - label * input
733:     File "/paddle/build/python/paddle/tensor/ops.py", line 580, in exp
733:       helper.append_op(type='exp', inputs={"X": x}, outputs={"Out": out})
733:     File "/paddle/build/python/paddle/fluid/layer_helper.py", line 45, in append_op
733:       return self.main_program.current_block().append_op(*args, **kwargs)
733:     File "/paddle/build/python/paddle/fluid/framework.py", line 4068, in append_op
733:       attrs=kwargs.get("attrs", None),
733:     File "/paddle/build/python/paddle/fluid/framework.py", line 2851, in __init__
733:       for frame in traceback.extract_stack():
733: 
733:     UnimplementedError: There are no kernels which are registered in the exp operator.
733:       [Hint: Expected kernels_iter != all_op_kernels.end(), but received kernels_iter == all_op_kernels.end().] (at /paddle/paddle/fluid/framework/operator.cc:2213)
733:       [operator < exp > error]
733: 
733: ----------------------------------------------------------------------
733: Ran 26 tests in 0.335s
733: 
733: FAILED (errors=1)
733: 
1/1 Test #733: test_poisson_nll_loss ............***Failed    3.26 sec

### Next Step
As we are unfamilar with the implementation of `exp` and `log`, we find this beyond our reach to support `float16` and `bfloat16` in `poisson_nll_loss` at current stage. If it is possible to merge this pr first and support `float16` and `bfloat16` when the support of these data types are fully tested in their unitests in the develop branch @luotao1 ?

--- a/python/paddle/fluid/tests/unittests/test_poisson_nll_loss.py
+++ b/python/paddle/fluid/tests/unittests/test_poisson_nll_loss.py
@@ -200,6 +200,20 @@ class TestPoissonNLLLossErrCase(TestPoissonNLLLossBasicCase):
         self.test_err_epsilon()
 
 
+class TestPoissonNLLLossFloat16Case(TestPoissonNLLLossBasicCase):
+    def test_api(self):
+        if core.is_compiled_with_cuda():
+            self.test_static_case(dtype="float16")
+            self.test_dynamic_case(dtype="float16")
+
+
+class TestPoissonNLLLossBfloat16Case(TestPoissonNLLLossBasicCase):
+    def test_api(self):
+        if core.is_compiled_with_cuda():
+            self.test_static_case(dtype="uint16")
+            self.test_dynamic_case(dtype="uint16")
+
+
 class TestPoissonNLLLossFloat32Case(TestPoissonNLLLossBasicCase):
     def test_api(self):
         self.test_static_case(dtype="float32")

--- a/python/paddle/nn/functional/loss.py
+++ b/python/paddle/nn/functional/loss.py
@@ -1331,10 +1331,16 @@ def l1_loss(input, label, reduction='mean', name=None):
             return unreduced
     else:
         check_variable_and_dtype(
-            input, 'input', ['float32', 'float64', 'int32', 'int64'], 'l1_loss'
+            input,
+            'input',
+            ['float32', 'float64', 'int32', 'int64'],
+            'l1_loss',
         )
         check_variable_and_dtype(
-            label, 'label', ['float32', 'float64', 'int32', 'int64'], 'l1_loss'
+            label,
+            'label',
+            ['float32', 'float64', 'int32', 'int64'],
+            'l1_loss',
         )
 
         if reduction == 'sum':
@@ -1487,11 +1493,11 @@ def poisson_nll_loss(
          input (Tensor):
             Input tensor, expectation of underlying Poisson distribution.
             The shape of input tensor should be `(N, *)` or `(*)` where `(*)` denotes any number of extra dimensions.
-            It's data type should be float32, float64.
+            It's data type should be float16, bfloat16, float32, float64.
          label (Tensor):
             Label tensor, random sampled from Poisson distribution :math:`label \sim \text{Poisson}(input)`.
             The shape of input tensor should be `(N, *)` or `(*)`, same shape as the input tensor.
-            It's data type should be float32, float64.
+            It's data type should be float16, bfloat16, float32, float64.
          log_input (bool, optional):
             Whether to the treat input tensor as log input.
             If ``True`` the loss is computed as, :math:`\exp(\text{input}) - \text{label} * \text{input}` .
@@ -1544,13 +1550,13 @@ def poisson_nll_loss(
     check_variable_and_dtype(
         input,
         'input',
-        ['float32', 'float64'],
+        ['float16', 'uint16', 'float32', 'float64'],
         'poisson_nll_loss',
     )
     check_variable_and_dtype(
         label,
         'label',
-        ['float32', 'float64'],
+        ['float16', 'uint16', 'float32', 'float64'],
         'poisson_nll_loss',
     )
 

--- a/python/paddle/tensor/math.py
+++ b/python/paddle/tensor/math.py
@@ -159,7 +159,7 @@ def log(x, name=None):
         return _C_ops.log(x)
     else:
         check_variable_and_dtype(
-            x, 'x', ['float16', 'float32', 'float64'], "log"
+            x, 'x', ['uint16', 'float16', 'float32', 'float64'], "log"
         )
         inputs = {'X': [x]}
         helper = LayerHelper('log', **locals())

--- a/python/paddle/tensor/ops.py
+++ b/python/paddle/tensor/ops.py
@@ -565,6 +565,7 @@ def exp(x, name=None):
             [
                 'int32',
                 'int64',
+                'uint16',
                 'float16',
                 'float32',
                 'float64',

